### PR TITLE
Feat/standard search

### PR DESCRIPTION
**Pull request by mingard**
## Description

This PR introduces a search endpoint and indexing routine.

To enable indexing on a field, you must specify a `search` block in the schema, including the weight.

```json
{
  "fields": {
    "title": {
      "type": "String",
      "label": "Title",
      "search": {
        "weight": 2
      }
  }
}
```

A search block must exist in the env config:
```json
"search": {
  "minLength": 3,
  "datastore": "@dadi/api-mongodb",
  "database": "search"
}
```

### minLength
The number of characters required before a search can be executed

### datastore
The selected database method

### database
The database to store words and search instances

## Running an initial index
To run a batch index against the entire dataset, run `node node_modules/@dadi/api/utils/index-search`. This may take some time, so have a ☕️ to hand.

## Running a query

A search endpoint has been introduced. It lives at collection level, and can be accessed by including a `q` param, which must be a basic string e.g.
> `https://somedomain.com/1.0/my-db/books/search?q=harry wizard`

Field filters can be applied in the same way as collection filtering.
> `https://somedomain.com/1.0/my-db/books/search?q=harry wizard&fields={"title": 1}`

### CODE REVIEW

**Comment by mingard** [comment]
@jimlambie I _believe_ this is ready to test and merge. I have privately shared a branch of a project where a single collection schema has the required parameters to work with search, and will provide some example queries to you directly, as well as information on how to initiate the indexing routine.

@fra967 This may be of interest.

File: dadi/lib/model/index.js [hunk 1]
```diff
@@ -679,6 +690,46 @@ Model.prototype.find = function (query, options, done) {
...
+    return done({
+      err: 'Search must be enabled and use a compatible datastore'
+    })
+  }
+  if (typeof options === 'function') {
+    done = options
+    options = {}
+  }
+  if (!options.search || options.search.length < config.get('search.minLength')) {
+    return done({
```
Comment (eduardoboucas) [inline dadi/lib/model/index.js]: (Same as above)

File: utils/index-search.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  const fields = help.getFieldsFromSchema(options.schema)
+  const database = options.database
+  const settings = options.schema.settings
+  const model = Model(options.name, JSON.parse(fields), null, settings, database)
+  this.collections[options.name] = model
+}
+
+BatchIndex.prototype.startCollectionImport = function () {
+  Object.keys(this.collections)
+    .map(name => {
```
Comment (eduardoboucas) [inline utils/index-search.js]: Curious as to why you're using `map` if you're not doing anything with the result. Do you mean `forEach`? You're also creating `const search` and not doing anything with it – could you just call `new` directly?

File: dadi/lib/controller/index.js [hunk 3]
```diff
@@ -59,6 +59,10 @@ function prepareQueryOptions (options, modelSettings) {
     response.errors.push(new ApiError('Bad Request', 'Invalid Parameter', 'The `page` parameter must be greater than zero', 'Invalid Page Parameter Provided'))
   }
 
+  if (options.q) {
```
[the remark this answers is not in the document]
Comment (mingard) [reply dadi/lib/controller/index.js]: It is a standard convention, but suggestions are welcome :)

File: dadi/lib/model/index.js [hunk 4]
```diff
@@ -679,6 +690,46 @@ Model.prototype.find = function (query, options, done) {
...
+ * Lookup documents in the database based on search query and run any associated hooks
+ *
+ * @param {Object} query
+ * @param {Function} done
+ * @return undefined
+ * @api public
+ */
+Model.prototype.search = function (options, done, req) {
+  if (!this.searcher.canUse()) {
+    return done({
```
[the remark this answers is not in the document]
Comment (mingard) [reply dadi/lib/model/index.js]: 👍

File: dadi/lib/model/index.js (hunk 1 above)
Comment (mingard) [reply dadi/lib/model/index.js]: 👍

File: utils/index-search.js (hunk 2 above)
Comment (mingard) [reply utils/index-search.js]: I was previously returning a mutation but not anymore. I'll change this now, and remove unnecessary constant.

File: dadi/lib/controller/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (eduardoboucas) [reply dadi/lib/controller/index.js]: It's representing the search terms, right? Maybe we keep `q` if it's a convention consumers are expecting and a simple comment in the code helps to make it clear what the variable contains?

[the remark this answers is not in the document]
Comment (mingard) [reply dadi/lib/controller/index.js]: That's right. I'll add a comment now.

**Comment by mingard** [comment]
@eduardoboucas take a look at the changes following your review. Have you been able to test the branch?

**Comment by eduardoboucas** [comment]
> @eduardoboucas take a look at the changes following your review. Have you been able to test the branch?

Looks good to me. No, I haven't had the chance to test yet. Will do next week.

> To run a batch index against the entire dataset, run node node_modules/@dadi/api/utils/index-search.

Just making a note for future reference that this should be a good candidate for [CLI](https://github.com/dadi/cli).

**Comment by mingard** [comment]
@eduardoboucas wonderful! 

> The only thing we could be missing is some acceptance tests

I'll start work on these as soon as we're happy with the approach. @jimlambie have you managed to take this for a spin? 🚗

**Pull request closed without merging**